src/main/java/pro/trevor/tankgame/util/Util.java

```diff
@@ -77,24 +129,58 @@ public static Set<Position> getSpacesInRange(Board board, Position p, int range)
         return output;
     }
 
-    public static boolean canMoveTo(State state, Position s, Position e) {
-        Position[] adjacent = allAdjacentPositions(s);
-        if (!Arrays.stream(adjacent).toList().contains(e)) {
-            return false;
-        } else if (!state.getBoard().isWalkable(e)) {
-            return false;
+    public static Set<Position> allPossibleMoves(Board board, Position p, int speed) {
+        return allPossiblePaths(board, p, speed).keySet();
+    }
+
+    /**
+     * Finds the shortest path to each space within the unit's speed. The result of this is stored into the return
+     * value as a hashmap of Position(Target) -> Pair[Integer(Remaining speed), Position(Previous position)].
+     * Ties in optimal paths are arbitrarily broken.
+     * @param board The board to check for if positions are valid.
+     * @param p The position to check for paths from.
+     * @param speed The number of spaces to search. Must be non-negative.
+     */
+    public static HashMap<Position, Pair<Integer, Position>> allPossiblePaths(Board board, Position p, int speed) {
```

Comment:
Please use full variable names (not p, same comment on other methods).  I appreciate the documentation.  

The use case for getting the whole path is for move animations.  It's probably better to keep this methods private for now (it exposes a lot of details) and add a method that constructs the single path that the player took when we need it.